File: heapSort.cpp
// HeapSort.cpp : 定义控制台应用程序的入口点。
//

#include "stdafx.h"
#include<iostream>
using namespace std;
#include "HeapSort.h"

void main() 
{
	HeapSort<int> *heapSort=new HeapSort<int>(20);
	heapSort->data[0]=6;  
	heapSort->data[1]=16;  
	heapSort->data[2]=15; 
	heapSort->data[3]=1;  
	heapSort->data[4]=12; 
	heapSort->data[5]=13; 
	heapSort->data[6]=2;  
	heapSort->data[7]=9;  
	heapSort->data[8]=10;  
	heapSort->data[9]=5;  
	heapSort->data[10]=3; 
	heapSort->data[11]=7; 
	heapSort->data[12]=10;
	heapSort->data[13]=11;
	heapSort->data[14]=14;
	heapSort->data[15]=17; 
	heapSort->data[16]=2; 
	heapSort->data[17]=18;
	heapSort->data[18]=20;
	heapSort->data[19]=19;
	heapSort->Sort();
	heapSort->Print();
}

File: src/heapSort.java
class HeapSort {
	private int[] data;
	private int length;

	public HeapSort(int len) {
		length = len;
		data = new int[length];
	}

	public HeapSort(int[] data) {
		super();
		this.data = data;
		length = data.length;
	}

	public void BuildHeap(int n) {
		for (int i = (n - 1) / 2; i >= 0; i--) {
			Heapify(i, n);
		}
	}

	public void Heapify(int low, int high) {
		int current = low;
		int child = 2 * current + 1;
		if (child <= high) {
			if (child + 1 < high) {
				if (data[child] < data[child + 1])
					child++;
			}
			if (data[current] < data[child]) {
				swap(current, child);
				Heapify(2 * current + 1, high);
				Heapify(2 * current + 2, high);
			}
		}
	}

	private void swap(int current, int child) {
		if (current != child) {
			data[child] ^= data[current];
			data[current] ^= data[child];
			data[child] ^= data[current];
		}
	}

	public void Print() {
		for (int i = 0; i < length; i++) {
			System.out.print(data[i] + ",");
		}
		System.out.println();
	}

	public void Print(int l, int r) {
		for (int i = l; i <= r; i++) {
			System.out.print(data[i] + ",");
		}
		System.out.println();
	}

	public void Sort() {
		for (int i = length - 1; i >= 1; i--) {
			BuildHeap(i);
			swap(0, i);
		}
	}
}
